File: vuln-sandbox/node/cwe-601/open-redirect-safe.js
// SAFE: open-redirect — redirect target validated against an allowlist
// Rule: ReactWindowLocation | CWE-601 | Expected: TrueNegative

const express = require('express');
const app = express();

const ALLOWED_REDIRECT_PATHS = ['/dashboard', '/profile', '/settings', '/home'];

app.get('/redirect', (req, res) => {
  const target = req.query.to;

  // SAFE: only allow redirects to known internal paths; reject external URLs
  if (!ALLOWED_REDIRECT_PATHS.includes(target)) {
    return res.redirect('/home');
  }

  res.redirect(target);
});

app.listen(3000);

File: vuln-sandbox/go/cwe-601/open-redirect-safe.go
// SAFE: go/http-redirect-user-input — redirect target validated against allowlist
// Rule: go/http-redirect-user-input | CWE-601 | Expected: TrueNegative

package main

import (
	"net/http"
)

var allowedPaths = map[string]bool{
	"/dashboard": true,
	"/profile":   true,
	"/settings":  true,
	"/home":      true,
}

func redirectHandler(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("to")

	// SAFE: only redirect to known internal paths; reject external URLs
	if !allowedPaths[target] {
		http.Redirect(w, r, "/home", http.StatusFound)
		return
	}

	http.Redirect(w, r, target, http.StatusFound)
}

func main() {
	http.HandleFunc("/redirect", redirectHandler)
	http.ListenAndServe(":8080", nil)
}
